### PR TITLE
[ValueTracking] Compute knownbits for `(and/or cond0, cond1)` on both sides of branch

### PR DESCRIPTION
**Pull request by goldsteinn**
The false branch for `and` and true branch for `or` provide less
information (intersection as opposed to union), but still can give
some useful information.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -711,10 +711,17 @@ static void computeKnownBitsFromCond(const Value *V, Value *Cond,
...
-    computeKnownBitsFromCond(V, A, Known, Depth + 1, SQ, Invert);
-    computeKnownBitsFromCond(V, B, Known, Depth + 1, SQ, Invert);
+      match(Cond, m_LogicalOp(m_Value(A), m_Value(B)))) {
+    KnownBits Known2(Known.getBitWidth());
+    KnownBits Known3(Known.getBitWidth());
+    computeKnownBitsFromCond(V, A, Known2, Depth + 1, SQ, Invert);
+    computeKnownBitsFromCond(V, B, Known3, Depth + 1, SQ, Invert);
+    if (Invert ? match(Cond, m_LogicalOr(m_Value(A), m_Value(B)))
+               : match(Cond, m_LogicalAnd(m_Value(A), m_Value(B))))
+      Known = Known2.unionWith(Known3);
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:721]:
```suggestion
      Known2 = Known2.unionWith(Known3);
```
You probably meant this?

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp]: Bah yup, Fixed.

**Comment by nikic** [comment]
@dtcxzyw Can you please rerun the tests?

@goldsteinn What's the compile-time impact?

**Comment by goldsteinn** [comment]
> @dtcxzyw Can you please rerun the tests?
> 
> @goldsteinn What's the compile-time impact?

No meaningful affect:
https://llvm-compile-time-tracker.com/compare.php?from=55bc0488af077acb47be70542718d1bc17f3de4f&to=a6a7dcd98c13617dfaf0f82614b06cc66a14694c&stat=instructions:u

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -711,10 +711,17 @@ static void computeKnownBitsFromCond(const Value *V, Value *Cond,
...
-      (Invert ? match(Cond, m_LogicalOr(m_Value(A), m_Value(B)))
-              : match(Cond, m_LogicalAnd(m_Value(A), m_Value(B))))) {
-    computeKnownBitsFromCond(V, A, Known, Depth + 1, SQ, Invert);
-    computeKnownBitsFromCond(V, B, Known, Depth + 1, SQ, Invert);
+      match(Cond, m_LogicalOp(m_Value(A), m_Value(B)))) {
+    KnownBits Known2(Known.getBitWidth());
+    KnownBits Known3(Known.getBitWidth());
+    computeKnownBitsFromCond(V, A, Known2, Depth + 1, SQ, Invert);
+    computeKnownBitsFromCond(V, B, Known3, Depth + 1, SQ, Invert);
+    if (Invert ? match(Cond, m_LogicalOr(m_Value(A), m_Value(B)))
```
Comment (topperc) [inline llvm/lib/Analysis/ValueTracking.cpp:719]: You can use m_Value() instead of m_Value(A/B). You don't need to recapture the variables.

**Review by nikic — CHANGES_REQUESTED** [review]
You'll also have to adjust the logic in DomCondCache to recurse into all and/or, not just ones that match the top-level type. Unless you intentionally only want to handle the inverted type at the top level?

Please recheck compile-time impact after doing that.

**Comment by goldsteinn** [comment]
> You'll also have to adjust the logic in DomCondCache to recurse into all and/or, not just ones that match the top-level type. Unless you intentionally only want to handle the inverted type at the top level?
> 
> Please recheck compile-time impact after doing that.

https://llvm-compile-time-tracker.com/compare.php?from=55bc0488af077acb47be70542718d1bc17f3de4f&to=d3e117a04afb14a00686eb689213776b2921154b&stat=instructions%3Au

Mostly minimal with two notable regressions:
Stage2-O0-g: tramp3d-v4	20259M	20310M (+0.25%)
Stage1-ReleaseLTO-g: consumer-typeset	58135M	58200M (+0.11%)

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**